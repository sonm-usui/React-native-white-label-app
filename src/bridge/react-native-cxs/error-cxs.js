// @flow
export const CONNECTION_ALREADY_EXISTS = '1062'
export const CREDENTIAL_DEFINITION_NOT_FOUND = '1036'
export const SCHEMA_NOT_FOUND = '1040'
export const INVALID_CREDENTIAL_OFFER = '1043'
export const ACTION_IS_NOT_SUPPORTED = '1103'
export const CLOUD_AGENT_UNAVAILABLE = '1010'
export const NO_POOL_OPEN = '1030'
export const CREDENTIAL_SCHEMA_NOT_FOUND = '1031'
export const WALLET_ITEM_NOT_FOUND = '1073'
export const INVALID_AGENCY_RESPONSE = '1020'
export const NOT_READY = '1005'
export const CONNECTION_UPGRADE_NOT_NEEDED = '1065'

export const CONNECTION_ALREADY_EXISTS_MESSAGE =
  'The connection invitation has already been accepted. ' +
  'Please, request a new invitation.'

export const CREDENTIAL_DEFINITION_NOT_FOUND_MESSAGE =
  'Definition for offered Credential not found on the network application is connected to.'

export const SCHEMA_NOT_FOUND_MESSAGE =
  'Schema for offered Credential not found on the network application is connected to.'

export const INVALID_CREDENTIAL_OFFER_MESSAGE =
  'Offered Credential does not match to its schema definition.'
