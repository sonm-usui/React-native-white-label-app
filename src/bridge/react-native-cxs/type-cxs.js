// @flow

import type { UserOneTimeInfo } from '../../store/user/type-user-store'
import type { AgencyPoolConfig } from '../../store/type-config-store'
import type { ClaimOfferPushPayload } from '../../push-notification/type-push-notification'
import type { ProprietaryConnectionInvitation } from '../../invitation/type-invitation'

export type AriesProtocolConfig = {
  protocol_type: string,
}

export type VcxProvision = {
  agency_url: string,
  agency_did: string,
  agency_verkey: string,
} & AriesProtocolConfig

export type VcxProvisionResult = {
  wallet_name: string,
  wallet_key: string,
  agency_endpoint: string,
  agency_did: string,
  agency_verkey: string,
  // myOneTimeDid
  sdk_to_remote_did: string,
  // myOneTimeVerificationKey
  sdk_to_remote_verkey: string,
  // oneTimeAgencyDid
  institution_did: string,
  // oneTimeAgencyVerificationKey
  institution_verkey: string,
  // myOneTimeAgentDid
  remote_to_sdk_did: string,
  // myOneTimeAgentVerificationKey
  remote_to_sdk_verkey: string,
}

export type CxsInitConfig = UserOneTimeInfo & AgencyPoolConfig

export type VcxInitConfig = {
  agency_endpoint: string,
  agency_did: string,
  agency_verkey: string,
  wallet_key: string,
  wallet_name: string,
  remote_to_sdk_did: string,
  remote_to_sdk_verkey: string,
  sdk_to_remote_did: string,
  sdk_to_remote_verkey: string,
  institution_did: string,
  institution_verkey: string,
} & AriesProtocolConfig

export type VcxPushTokenConfig = {
  id: string,
  value: string,
}

export type CxsPushTokenConfig = {
  uniqueId: string,
  pushToken: string,
}

export type CxsPoolConfig = {
  poolConfig: string,
}

export type CxsPoolConfigWithGenesisPath = CxsPoolConfig & {
  genesis_path: string,
}

export type VcxPoolInitConfig = {
  genesis_path: string,
  pool_name?: string,
}

export type VcxCreateConnection = {
  source_id: string,
  invite_details: ProprietaryConnectionInvitation,
}

export type VcxConnectionConnectResult = {|
  source_id: string,
  pw_did: string,
  pw_verkey: string,
  uuid: string,
  endpoint: string,
  agent_did: string,
  agent_vk: string,
  their_pw_did: string,
  their_pw_verkey: string,
|}

export type VcxCredentialOffer = {
  msg_type: string,
  version: string,
  to_did: string,
  from_did: string,
  libindy_offer: string,
  cred_def_id: string,
  credential_attrs: { [string]: Array<string> },
  schema_seq_no: number,
  claim_name: string,
  claim_id: string,
  payment_address?: ?string,
  price?: ?string,
}

export type CxsCredentialOfferResult = {
  claimHandle: number,
  claimOffer: ClaimOfferPushPayload,
}

export type VcxClaimInfo = {
  credential_id?: string,
  credential?: string,
  credential_offer?: string,
}

export type WalletPoolName = {
  walletName: string,
  poolName: string,
}

export type UTXO = {
  amount: number,
  extra: string,
  paymentAddress: string,
  source: string,
}

export type PaymentAddress = {
  address: string,
  balance: number,
  utxo: UTXO[],
}

export type WalletTokenInfo = {
  balance_str: string,
  addresses: PaymentAddress[],
}

export const smallDeviceMemory = 1073741824

export type SignDataResponse = {
  data: string,
  signature: string,
}

export const signDataResponseSchema = {
  type: 'object',
  properties: {
    data: { type: 'string', minLength: 1 },
    signature: { type: 'string', minLength: 1 },
  },
  required: ['data', 'signature'],
}
