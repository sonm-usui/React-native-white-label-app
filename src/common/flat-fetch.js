// @flow

// If you look into the return type of this function
// you should notice that we are returning an array from  this function
// this function is just a wrapper on top of fetch.
// However, this changes a lot of ways we write code and handle errors
// By returning an array in which first item is an error
// and second item is the response
// We get two important things
// 1. We don't have to write try...catch statements everywhere
// 2. We are forced to handle errors. No more forgetting to add try...catch
// TODO:KS Write more details about both 1 & 2
// Problem #1 leads to try...catch nesting hell if there are multiple async
// calls
// Problem #2 leads to unexpected app crashes at runtime

export const acceptHeaders = { Accept: 'application/json;flavor=didcomm-msg' }

export async function flatFetch(
  url: string,
  body?: string,
  headers?: { [string]: any }
): Promise<[null | typeof Error, null | string, null | Response]> {
  const options = {
    mode: 'cors',
    ...(body ? { method: 'POST', body } : {}),
    ...(headers ? headers : {}),
  }
  try {
    const response = await fetch(url, options)
    const responseText = await response.text()
    // request redirected
    if (response && response.status === 302) {
      return [null, null, response]
    }
    if (!response.ok) {
      throw new Error(responseText)
    }
    return [null, responseText, response]
  } catch (e) {
    return [e, null, null]
  }
}
