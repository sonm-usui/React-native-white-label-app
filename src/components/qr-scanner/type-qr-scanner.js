// @flow

import type {
  AriesConnectionInvite,
  AriesOutOfBandInvite,
  InvitationPayload,
  ShortProprietaryConnectionInvitation,
} from '../../invitation/type-invitation'
import type { QrCodeEphemeralProofRequest } from '../../proof-request/type-proof-request'
import type { QrCodeEphemeralCredentialOffer } from '../../claim-offer/ephemeral-claim-offer'

export const SCAN_STATUS = {
  SCANNING: 'scanning...',
  SUCCESS: 'Success!',
  FAIL: 'QR code format is invalid.',
  DOWNLOADING_INVITATION: 'Downloading invitation...',
  NO_INVITATION_DATA: 'No message found behind this QR code.',
  DOWNLOADING_AUTHENTICATION_JWT: 'Loading authentication request...',
  NO_AUTHENTICATION_REQUEST:
    'No authentication request found behind this QR code.',
  AUTH_REQUEST_DOWNLOAD_FAILED: 'Failed to load authentication request.',
  AUTH_REQUEST_INVALID_HEADER_DECODE_ERROR:
    '001::Malformed authentication request.',
  AUTH_REQUEST_INVALID_HEADER_SCHEMA: '002::Malformed authentication request.',
  AUTH_REQUEST_INVALID_BODY_DECODE_ERROR:
    '003::Malformed authentication request.',
  AUTH_REQUEST_INVALID_BODY_SCHEMA: '004::Malformed authentication request.',
  AUTH_REQUEST_INVALID_SIGNATURE: '005::Malformed authentication request.',
  AUTH_REQUEST_INVALID_BODY_SCHEMA_AND_SEND_FAIL:
    '006::Malformed authentication request.',
  DOWNLOADING: 'Downloading...',
  INVALID_DOWNLOADED_DATA: '007::Invalid data from QR Code.',
  INVALID_URL_QR_CODE: '008::Invalid QR code.',
  INVALID_OPENID_QR_LINK: '009::Invalid OpenID link.',
}

export type QR_SCAN_STATUS =
  | typeof SCAN_STATUS.SCANNING
  | typeof SCAN_STATUS.FAIL
  | typeof SCAN_STATUS.SUCCESS
  | typeof SCAN_STATUS.DOWNLOADING_INVITATION
  | typeof SCAN_STATUS.NO_INVITATION_DATA
  | typeof SCAN_STATUS.DOWNLOADING_AUTHENTICATION_JWT
  | typeof SCAN_STATUS.NO_AUTHENTICATION_REQUEST
  | typeof SCAN_STATUS.AUTH_REQUEST_DOWNLOAD_FAILED
  | typeof SCAN_STATUS.AUTH_REQUEST_INVALID_HEADER_DECODE_ERROR
  | typeof SCAN_STATUS.AUTH_REQUEST_INVALID_HEADER_SCHEMA
  | typeof SCAN_STATUS.AUTH_REQUEST_INVALID_BODY_DECODE_ERROR
  | typeof SCAN_STATUS.AUTH_REQUEST_INVALID_BODY_SCHEMA
  | typeof SCAN_STATUS.AUTH_REQUEST_INVALID_SIGNATURE
  | typeof SCAN_STATUS.AUTH_REQUEST_INVALID_BODY_SCHEMA_AND_SEND_FAIL
  | typeof SCAN_STATUS.DOWNLOADING
  | typeof SCAN_STATUS.INVALID_DOWNLOADED_DATA
  | typeof SCAN_STATUS.INVALID_URL_QR_CODE

export type QrScannerState = {
  scanStatus: QR_SCAN_STATUS,
}

export type QrScannerProps = {
  onShortProprietaryInvitationRead: (ShortProprietaryConnectionInvitation) => Promise<void>,
  onClose: () => void,
  onProprietaryInvitationRead: (InvitationPayload) => Promise<void>,
  onOIDCAuthenticationRequest: (OIDCAuthenticationRequest) => void,
  onAriesConnectionInviteRead: (AriesConnectionInvite) => Promise<void>,
  onAriesOutOfBandInviteRead: (AriesOutOfBandInvite) => Promise<void>,
  onEphemeralProofRequest: (QrCodeEphemeralProofRequest) => void,
  onEphemeralCredentialOffer: (QrCodeEphemeralCredentialOffer) => Promise<void>,
}

export type CameraMarkerProps = {
  status: QR_SCAN_STATUS,
  onClose: () => void,
}

export const TOP_LEFT = 'topLeft'
export const TOP_RIGHT = 'topRight'
export const BOTTOM_LEFT = 'bottomLeft'
export const BOTTOM_RIGHT = 'bottomRight'

export type CornerBoxProps = {
  status: QR_SCAN_STATUS,
  position:
    | typeof TOP_LEFT
    | typeof TOP_RIGHT
    | typeof BOTTOM_LEFT
    | typeof BOTTOM_RIGHT,
}

export const QR_CODE_TYPES = {
  INVITATION: 'INVITATION',
  URL_INVITATION: 'URL_INVITATION',
  ENV_SWITCH_URL: 'ENV_SWITCH_URL',
  OIDC: 'OIDC',
  EPHEMERAL_PROOF_REQUEST_V1: 'EPHEMERAL_PROOF_REQUEST_V1',
  // this is the type of QR code, where QR code is a url, and response from URL is not a json object, instead it is non-json string
  // it could be a base64 encoded data, so we need to keep this type
  // so that other types of QR handler can handle such type of qr codes
  URL_NON_JSON_RESPONSE: 'URL_NON_JSON_RESPONSE',
  OUTOFBAND_PROOF_REQUEST: 'OUTOFBAND_PROOF_REQUEST',
  EPHEMERAL_CREDENTIAL_OFFER: 'EPHEMERAL_CREDENTIAL_OFFER',
}
export type QrCodeTypes = $Keys<typeof QR_CODE_TYPES>

export type QrCodeOIDC = {
  type: 'OIDC',
  version: string,
  clientId: string,
  requestUri: string,
  responseType: string,
}

export type JWTAuthenticationRequest = {
  header: {
    alg: string,
    typ: string,
    kid: ?string,
  },
  body: {
    iss: string,
    response_type: string,
    client_id: string,
    scope: string,
    state: string,
    nonce: string,
    response_mode: string,
    registration: {
      request_object_signing_alg: ?string,
      jwks_uri: ?string,
      id_token_signed_response_alg: Array<string>,
    },
  },
  encodedSignature: ?string,
}

export type OIDCAuthenticationRequest = {
  oidcAuthenticationQrCode: QrCodeOIDC,
  jwtAuthenticationRequest: JWTAuthenticationRequest,
  id: string,
}

export type QrCodeNonJsonUrl = {
  type: 'URL_NON_JSON_RESPONSE',
  data: string,
}
