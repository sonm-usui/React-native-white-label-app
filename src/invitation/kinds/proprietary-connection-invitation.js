import {
  QR_CODE_DELEGATION_DID,
  QR_CODE_DELEGATION_KEY,
  QR_CODE_DELEGATION_SIGNATURE,
  QR_CODE_LOGO_URL,
  QR_CODE_REQUEST_ID,
  QR_CODE_SENDER_AGENCY,
  QR_CODE_SENDER_AGENCY_DID,
  QR_CODE_SENDER_AGENCY_ENDPOINT,
  QR_CODE_SENDER_AGENCY_KEY,
  QR_CODE_SENDER_DETAIL,
  QR_CODE_SENDER_DID,
  QR_CODE_SENDER_KEY_DELEGATION,
  QR_CODE_SENDER_NAME,
  QR_CODE_SENDER_PUBLIC_DID,
  QR_CODE_SENDER_VERIFICATION_KEY,
  QR_CODE_TARGET_NAME,
  QR_CODE_VERSION
} from "../../api/api-constants";
import {schemaValidator} from "../../services/schema-validator";
import type {InvitationPayload, ShortProprietaryConnectionInvitation, ProprietaryConnectionInvitation} from "../type-invitation";

export function isShortProprietaryInvitation(
  data: Object
): ShortProprietaryConnectionInvitation | false {
  // check if qr code schema is valid
  if (!schemaValidator.validate(proprietaryShortInviteSchema, data)) {
    return false
  }

  // since all checks are passed
  // qr code data is valid, send the parsed data back
  return data
}

export function isProprietaryInvitation(
  data: Object
): ProprietaryConnectionInvitation | false {
  // check if qr code schema is valid
  if (!schemaValidator.validate(proprietaryInvitationSchema, data)) {
    return false
  }

  // since all checks are passed
  // qr code data is valid, send the parsed data back
  return data
}

export function convertShortProprietaryInvitationToAppInvitation(qrCode: ShortProprietaryConnectionInvitation) {
  const qrSenderDetail = qrCode[QR_CODE_SENDER_DETAIL]
  const qrSenderAgency = qrCode[QR_CODE_SENDER_AGENCY]
  const senderDetail = {
    name: qrSenderDetail[QR_CODE_SENDER_NAME],
    agentKeyDlgProof: {
      agentDID:
        qrSenderDetail[QR_CODE_SENDER_KEY_DELEGATION][QR_CODE_DELEGATION_DID],
      agentDelegatedKey:
        qrSenderDetail[QR_CODE_SENDER_KEY_DELEGATION][QR_CODE_DELEGATION_KEY],
      signature:
        qrSenderDetail[QR_CODE_SENDER_KEY_DELEGATION][
          QR_CODE_DELEGATION_SIGNATURE
          ],
    },
    DID: qrSenderDetail[QR_CODE_SENDER_DID],
    logoUrl: qrSenderDetail[QR_CODE_LOGO_URL],
    verKey: qrSenderDetail[QR_CODE_SENDER_VERIFICATION_KEY],
    publicDID: qrSenderDetail[QR_CODE_SENDER_PUBLIC_DID],
  }

  const senderAgencyDetail = {
    DID: qrSenderAgency[QR_CODE_SENDER_AGENCY_DID],
    verKey: qrSenderAgency[QR_CODE_SENDER_AGENCY_KEY],
    endpoint: qrSenderAgency[QR_CODE_SENDER_AGENCY_ENDPOINT],
  }

  return {
    senderEndpoint: senderAgencyDetail.endpoint,
    requestId: qrCode[QR_CODE_REQUEST_ID],
    senderAgentKeyDelegationProof: senderDetail.agentKeyDlgProof,
    senderName: senderDetail.name,
    senderDID: senderDetail.DID,
    senderLogoUrl: senderDetail.logoUrl,
    senderVerificationKey: senderDetail.verKey,
    targetName: qrCode[QR_CODE_TARGET_NAME],
    senderDetail,
    senderAgencyDetail,
    version: qrCode[QR_CODE_VERSION],
  }
}

export const convertProprietaryInvitationToAppInvitation = (
  pendingInvitation: ProprietaryConnectionInvitation
): InvitationPayload => ({
  senderEndpoint: pendingInvitation.senderAgencyDetail.endpoint,
  requestId: pendingInvitation.connReqId,
  senderAgentKeyDelegationProof:
  pendingInvitation.senderDetail.agentKeyDlgProof,
  senderName: pendingInvitation.senderDetail.name,
  senderDID: pendingInvitation.senderDetail.DID,
  senderLogoUrl: pendingInvitation.senderDetail.logoUrl,
  senderVerificationKey: pendingInvitation.senderDetail.verKey,
  targetName: pendingInvitation.targetName,
  senderAgencyDetail: pendingInvitation.senderAgencyDetail,
  senderDetail: pendingInvitation.senderDetail,
  version: pendingInvitation.version,
})

export const proprietaryShortInviteSchema = {
  type: 'object',
  properties: {
    [QR_CODE_REQUEST_ID]: {type: 'string'},
    [QR_CODE_SENDER_DETAIL]: {
      type: 'object',
      properties: {
        [QR_CODE_SENDER_DID]: {type: 'string'},
        [QR_CODE_SENDER_NAME]: {type: 'string'},
        [QR_CODE_LOGO_URL]: {type: 'string'},
        [QR_CODE_SENDER_VERIFICATION_KEY]: {type: 'string'},
        [QR_CODE_SENDER_KEY_DELEGATION]: {
          type: 'object',
          properties: {
            [QR_CODE_DELEGATION_DID]: {type: 'string'},
            [QR_CODE_DELEGATION_KEY]: {type: 'string'},
            [QR_CODE_DELEGATION_SIGNATURE]: {type: 'string'},
          },
          required: [
            QR_CODE_DELEGATION_DID,
            QR_CODE_DELEGATION_KEY,
            QR_CODE_DELEGATION_SIGNATURE,
          ],
        },
      },
      required: [
        QR_CODE_SENDER_DID,
        QR_CODE_SENDER_NAME,
        QR_CODE_LOGO_URL,
        QR_CODE_SENDER_VERIFICATION_KEY,
        QR_CODE_SENDER_KEY_DELEGATION,
      ],
    },
    [QR_CODE_SENDER_AGENCY]: {
      type: 'object',
      properties: {
        [QR_CODE_SENDER_AGENCY_DID]: {type: 'string'},
        [QR_CODE_SENDER_AGENCY_KEY]: {type: 'string'},
        [QR_CODE_SENDER_AGENCY_ENDPOINT]: {type: 'string'},
      },
      required: [
        QR_CODE_SENDER_AGENCY_DID,
        QR_CODE_SENDER_AGENCY_KEY,
        QR_CODE_SENDER_AGENCY_ENDPOINT,
      ],
    },
    [QR_CODE_TARGET_NAME]: {type: 'string'},
  },
  required: [
    QR_CODE_REQUEST_ID,
    QR_CODE_SENDER_DETAIL,
    QR_CODE_SENDER_AGENCY,
    QR_CODE_TARGET_NAME,
  ],
}

export const proprietaryInvitationSchema = {
  type: 'object',
  properties: {
    connReqId: { type: 'string' },
    statusCode: { type: 'string' },
    senderDetail: {
      type: 'object',
      properties: {
        name: { type: 'string' },
        agentKeyDlgProof: {
          type: 'object',
          properties: {
            agentDID: { type: 'string' },
            agentDelegatedKey: { type: 'string' },
            signature: { type: 'string' },
          },
          required: ['agentDID', 'agentDelegatedKey', 'signature'],
        },
        DID: { type: 'string' },
        logoUrl: { type: 'string' },
        verKey: { type: 'string' },
        publicDID: { type: 'string' },
      },
      required: ['name', 'agentKeyDlgProof', 'DID', 'logoUrl', 'verKey'],
    },
    senderAgencyDetail: {
      type: 'object',
      properties: {
        DID: { type: 'string' },
        verKey: { type: 'string' },
        endpoint: { type: 'string' },
      },
      required: [],
    },
    targetName: { type: 'string' },
    version: { type: 'string' },
  },
  required: ['senderDetail', 'senderAgencyDetail', 'targetName'],
}
