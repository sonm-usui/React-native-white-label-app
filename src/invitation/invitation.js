// @flow
import React, { Component } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import type { Store } from '../store/type-store'
import type { ResponseTypes } from '../components/request/type-request'
import type { InvitationProps, InvitationNavigation } from './type-invitation'

import { Container } from '../components'
import { homeDrawerRoute, homeRoute, invitationRoute } from '../common'
import { ResponseType } from '../components/request/type-request'
import { sendInvitationResponse, invitationRejected } from './invitation-store'
import { smsPendingInvitationSeen } from '../sms-pending-invitation/sms-pending-invitation-store'
import { SMSPendingInvitationStatus } from '../sms-pending-invitation/type-sms-pending-invitation'
import { Request } from '../components/request/request'
import { Platform } from 'react-native'
import { allowPushNotifications } from '../push-notification/push-notification-store'
import { usePushNotifications } from '../external-imports'

export class Invitation extends Component<InvitationProps, void> {
  render() {
    const { invitation } = this.props

    const {
      isValid,
      senderName,
      title,
      message,
      senderLogoUrl,
    } = isValidInvitation(this.props.invitation)

    if (!isValid) {
      return <Container />
    }

    return (
      <Container>
        <Request
          title={title}
          message={message}
          senderLogoUrl={senderLogoUrl}
          onAction={this.onAction}
          testID={'invitation'}
          invitationError={invitation ? invitation.error : undefined}
          senderName={senderName}
        />
      </Container>
    )
  }

  componentDidMount() {
    if (this.props.isSmsInvitationNotSeen) {
      this.props.smsPendingInvitationSeen(this.props.smsToken)
    }
  }

  hideModal = () => {
    const backRedirectRoute = this.props.route.params?.backRedirectRoute
    if (backRedirectRoute) {
      this.props.navigation.navigate(backRedirectRoute)
    } else {
      this.props.navigation.navigate(homeRoute, {
        screen: homeDrawerRoute,
        params: undefined,
      })
    }
  }

  onAction = (response: ResponseTypes) => {
    const { invitation } = this.props
    if (invitation) {
      const { payload } = invitation
      if (payload) {
        if (response === ResponseType.accepted) {
          if (Platform.OS === 'android' && usePushNotifications) {
            // for android, we don't need to show push permission screen
            // so we can safely register push notification on Android
            // For ios, we are already showing permission screen
            // and if user click on allow permission, then our saga would request
            // permission and update the FCM token
            this.props.allowPushNotifications()
          }
          this.props.sendInvitationResponse({
            response,
            senderDID: payload.senderDID,
          })
          this.props.navigation.navigate(homeRoute, {
            screen: homeDrawerRoute,
            params: undefined,
          })
        } else if (response === ResponseType.rejected) {
          this.props.invitationRejected(payload.senderDID)
          this.hideModal()
        }
      } else {
        this.hideModal()
      }
    }
  }
}

function isValidInvitation(
  invitation: $PropertyType<InvitationProps, 'invitation'>
) {
  let senderName = 'Anonymous'
  let title = 'Hi'
  let message = 'Anonymous wants to connect with you.'

  if (invitation && invitation.payload) {
    const { payload } = invitation
    senderName = payload.senderName || senderName
    message = `${senderName} wants to connect with you.`
    return {
      isValid: true,
      senderName,
      title,
      message,
      senderLogoUrl: payload.senderLogoUrl,
    }
  }

  return {
    isValid: false,
    senderName,
    title,
    message,
    senderLogoUrl: undefined,
  }
}

const mapStateToProps = (
  state: Store,
  { route: { params } }: InvitationNavigation
) => {
  const senderDID = params ? params.senderDID : ''
  const smsToken = params ? params.token : null
  const isSmsInvitationNotSeen =
    smsToken &&
    state.smsPendingInvitation[smsToken] &&
    state.smsPendingInvitation[smsToken].status !==
      SMSPendingInvitationStatus.SEEN

  return {
    invitation: state.invitation[senderDID],
    smsToken,
    isSmsInvitationNotSeen,
  }
}

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      sendInvitationResponse,
      invitationRejected,
      smsPendingInvitationSeen,
      allowPushNotifications,
    },
    dispatch
  )

export const invitationScreen = {
  routeName: invitationRoute,
  screen: connect(mapStateToProps, mapDispatchToProps)(Invitation),
}
