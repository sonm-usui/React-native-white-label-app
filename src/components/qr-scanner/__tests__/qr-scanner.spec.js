// @flow
import React from 'react'
import 'react-native'
import renderer from 'react-test-renderer'
import { SCAN_STATUS } from '../type-qr-scanner'
import QRScanner from '../qr-scanner'
import {
  qrData,
  validQrCodeEnvironmentSwitchUrl,
  validInvitationUrlQrCode,
  smsDownloadedPayload,
} from '../../../../__mocks__/static-data'
import { mockAriesV1QrCode } from '../../../../__mocks__/data/mock-qr-data'
import * as fetch from '../../../common/flat-fetch'
import { CONNECTION_INVITE_TYPES } from '../../../invitation/type-invitation'
import * as vcx from '../../../bridge/react-native-cxs/RNCxs'
import {convertProprietaryInvitationToAppInvitation} from "../../../invitation/kinds/proprietary-connection-invitation";

describe('<QRScanner />', () => {
  const getProps = () => ({
    onClose: jest.fn(),
    onRead: jest.fn(),
    onShortProprietaryInvitationRead: jest.fn(),
    onEnvironmentSwitchUrl: jest.fn(),
    onProprietaryInvitationRead: jest.fn(),
    onOIDCAuthenticationRequest: jest.fn(),
    onAriesConnectionInviteRead: jest.fn(),
    onAriesOutOfBandInviteRead: jest.fn(),
    onEphemeralProofRequest: jest.fn(),
    onEphemeralCredentialOffer: jest.fn(),
  })

  function setup() {
    const props = getProps()
    const wrapper = renderer.create(<QRScanner {...props} />)
    const instance = wrapper.getInstance()
    return {
      ...props,
      instance,
      wrapper,
    }
  }

  it('should match snapshot', () => {
    const { wrapper } = setup()
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  it('should call onRead once QR code read is successful', async () => {
    jest.useFakeTimers()
    const { onShortProprietaryInvitationRead, instance } = setup()

    const qrReadEvent = {
      data: JSON.stringify(qrData),
    }

    await instance.onRead(qrReadEvent)
    expect(onShortProprietaryInvitationRead).toHaveBeenCalledWith(expect.objectContaining(qrData))
    expect(instance.state.scanStatus).toBe(SCAN_STATUS.SCANNING)
  })

  it('should set state to fail if QR code is not correct', async () => {
    jest.useFakeTimers()

    const { instance } = setup()

    await instance.onRead({ data: '' })
    expect(instance.state.scanStatus).toBe(SCAN_STATUS.FAIL)

    jest.runAllTimers()

    expect(instance.state.scanStatus).toBe(SCAN_STATUS.SCANNING)
  })

  // this functionality is commented as of now
  xit('should call onEnvironmentSwitchUrl if it reads correct environment switcher url', async () => {
    jest.useFakeTimers()
    const { onEnvironmentSwitchUrl, instance } = setup()

    await instance.onRead({ data: validQrCodeEnvironmentSwitchUrl })
    expect(onEnvironmentSwitchUrl).toHaveBeenCalledWith({
      name: 'dev',
      url: validQrCodeEnvironmentSwitchUrl,
    })
    expect(instance.state.scanStatus).toBe(SCAN_STATUS.SUCCESS)
  })

  it('should send a request to download invitation if url is scanned', async () => {
    jest.useFakeTimers()
    const invitationDetailRequestSpy = jest.spyOn(fetch, 'flatFetch')

    invitationDetailRequestSpy.mockImplementation(() =>
      Promise.resolve([null, JSON.stringify(smsDownloadedPayload)])
    )

    const { onProprietaryInvitationRead, instance } = setup()

    const pendingQrProcessing = instance.onRead({
      data: validInvitationUrlQrCode,
    })
    expect(instance.state.scanStatus).toBe(SCAN_STATUS.DOWNLOADING)
    // process API call
    await pendingQrProcessing

    expect(onProprietaryInvitationRead).toHaveBeenCalledWith(
      convertProprietaryInvitationToAppInvitation(smsDownloadedPayload)
    )
    expect(instance.state.scanStatus).toBe(SCAN_STATUS.SCANNING)

    invitationDetailRequestSpy.mockReset()
    invitationDetailRequestSpy.mockRestore()

    jest.runAllTimers()

    expect(instance.state.scanStatus).toBe(SCAN_STATUS.SCANNING)
  })

  it('should call onAriesConnectionInviteRead, if we get aries QR code', async () => {
    jest.useFakeTimers()
    const { onAriesConnectionInviteRead, instance } = setup()

    const qrReadEvent = {
      data: JSON.stringify(mockAriesV1QrCode),
    }

    await instance.onRead(qrReadEvent)
    expect(onAriesConnectionInviteRead).toHaveBeenCalledWith(
      expect.objectContaining({
        payload: mockAriesV1QrCode,
        type: CONNECTION_INVITE_TYPES.ARIES_V1_QR,
        version: '1.0',
      })
    )
    expect(instance.state.scanStatus).toBe(SCAN_STATUS.SCANNING)
  })

  it('should call onAriesConnectionInviteRead, if we pass a url that gives aries v1 qr code', async () => {
    jest.useFakeTimers()

    const toUtf8Spy = jest.spyOn(vcx, 'toUtf8FromBase64')
    toUtf8Spy.mockImplementation(() =>
      Promise.resolve(JSON.stringify(mockAriesV1QrCode))
    )

    const { onAriesConnectionInviteRead, instance } = setup()

    const pendingQrProcessing = instance.onRead({
      data: `${validInvitationUrlQrCode}?c_i=${JSON.stringify(
        mockAriesV1QrCode
      )}`,
    })
    expect(instance.state.scanStatus).toBe(SCAN_STATUS.DOWNLOADING)
    // process API call
    await pendingQrProcessing

    expect(onAriesConnectionInviteRead).toHaveBeenCalledWith(
      expect.objectContaining({
        payload: mockAriesV1QrCode,
        type: CONNECTION_INVITE_TYPES.ARIES_V1_QR,
        version: '1.0',
      })
    )
    expect(instance.state.scanStatus).toBe(SCAN_STATUS.SCANNING)

    toUtf8Spy.mockReset()
    toUtf8Spy.mockRestore()

    jest.runAllTimers()

    expect(instance.state.scanStatus).toBe(SCAN_STATUS.SCANNING)
  })
})
