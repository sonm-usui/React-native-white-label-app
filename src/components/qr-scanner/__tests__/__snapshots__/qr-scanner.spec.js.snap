// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QRScanner /> should match snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;
